README.md
# clue/wol-react [![Build Status](https://travis-ci.org/clue/wol-react.svg?branch=master)](https://travis-ci.org/clue/wol-react)

Turn on your PC with [Wake-On-LAN](http://en.wikipedia.org/wiki/Wake-on-LAN) (WOL) requests

> Note: This project is in beta stage! Feel free to report any issues you encounter.

## Usage

Once [installed](#install), using this library is as simple as running:

```php
$loop = React\EventLoop\Factory::create();
$wolFactory = new Clue\React\Wol\Factory($loop);

$wolFactory->createSender()->then(function(Clue\React\Wol\Sender $wol) {
    $wol->send('11:22:33:44:55:66');
});

$loop->run();
```

If your environment requires a non-default broadcast address (the default is: `255.255.255.255:7`), for example `1.2.3.4:9`. You can pass that to the `createSender` method on the `Factory`. For example:


```php
$loop = React\EventLoop\Factory::create();
$wolFactory = new Clue\React\Wol\Factory($loop);

$wolFactory->createSender('1.2.3.4:9')->then(function(Clue\React\Wol\Sender $wol) {
    $wol->send('11:22:33:44:55:66');
});

$loop->run();
```

There's also a CLI script in `bin/wol.php` to send a WOL request from the 
command line simply by running:

```bash
$ php bin/wol.php 11:22:33:44:55:66
```

## Introduction

The following short introduction is mostly taken from wikipedia's
[article about WOL](http://en.wikipedia.org/wiki/Wake-on-LAN):

Wake-on-LAN ("WOL") is implemented using a specially designed packet called a
magic packet, which is sent to the computer to be woken up. The magic packet
contains the MAC address of the destination computer. Powered-down computers
capable of Wake-on-LAN will contain network devices able to "listen" to incoming
packets in low-power mode while the system is powered down. If a magic packet is
received that is directed to the device's MAC address, the NIC signals the
computer's power supply to initiate system wake-up, much in the same way as
pressing the power button would do.

The magic packet is a broadcast frame containing anywhere within its payload 6
bytes of all 255 (FF FF FF FF FF FF in hexadecimal), followed by sixteen
repetitions of the target computer's 48-bit MAC address, for a total of 102
bytes.

Since the magic packet is only scanned for the string above, and not actually
parsed by a full protocol stack, it may be sent as any network- and
transport-layer protocol, although this library uses a typical UDP datagram.
The magic packet is usually sent on the data link layer (layer 2 in the OSI
model) and when sent, is broadcast to all attached devices on a given network,
using the network broadcast address; the IP-address (layer 3 in the OSI model)
is not used.

## Install

The recommended way to install this library is [through composer](http://getcomposer.org). [New to composer?](http://getcomposer.org/doc/00-intro.md)

```JSON
{
    "require": {
        "clue/wol-react": "0.1.*"
    }
}
```

## License

MIT



CHANGELOG.md
# CHANGELOG

This file is a manually maintained list of changes for each release. Feel free
to add your changes here when sending pull requests. Also send corrections if
you spot any mistakes.

## 0.1.0 (2014-05-10)

* First tagged release
* Refactor construction to support custom broadcast addresses
  ([#2](https://github.com/clue/reactphp-wake-on-lan/pull/2) @WyriHaximus)
* Use stable/tagged dependencies
  ([#2](https://github.com/clue/reactphp-wake-on-lan/pull/2))
* Use `Clue\React\Wol` namespace and PSR-4 code layout
* Extended test suite with 100% coverage and continuos integration by Travis CI

## 0.0.0 (2013-04-13)

* Initial concept


tests/SenderTest.php
<?php

use Clue\React\Wol\Sender;

class SenderTest extends TestCase
{
    public function setUp()
    {
        $this->socket = $this->getMockBuilder('Socket\React\Datagram\Socket')
                             ->disableOriginalConstructor()
                             ->getMock();

        $this->sender = new Sender($this->socket);
    }

    public function testSend()
    {
        $this->socket
             ->expects($this->once())
             ->method('send');

        $this->sender->send('aa:bb:cc:dd:ee:ff');
    }

    /**
     * @dataProvider provideValidAddress
     * @param string $address
     * @param string $mac
     */
    public function testValidAddress($address, $mac)
    {
        $this->assertEquals($mac, $this->sender->coerceMac($address));
    }

    public function provideValidAddress()
    {
        return array(
            array(
                'Aa:bB:12:EF:34:56',
                'AA:BB:12:EF:34:56'
            ),
            array(
                'Aa-bB-12-EF-34-56',
                'AA:BB:12:EF:34:56'
            ),
            array(
                'AabB12EF3456',
                'AA:BB:12:EF:34:56'
            ),
        );
    }

    /**
     * @dataProvider provideInvalidAddress
     * @expectedException InvalidArgumentException
     * @param unknown $address
     */
    public function testInvalidAddress($address)
    {
       $this->sender->coerceMac($address);
    }

    public function provideInvalidAddress($address)
    {
        return array(
            array('aa:bb:cc'),
            array('aa:aa:aa:aa:aa:aa:aa'),
            array('gg:gg:gg:gg:gg:gg'),
        );
    }
}


tests/FactoryTest.php
<?php

use Clue\React\Wol\Factory;
use Clue\React\Wol\Sender;

class FactoryTest extends TestCase
{
    public function setUp()
    {
        $this->loop = React\EventLoop\Factory::create();
        $this->factory = new Factory($this->loop);
    }

    public function testSender()
    {
        $ret = $this->factory->createSender();

        $this->assertInstanceOf('React\Promise\PromiseInterface', $ret);

        $ret->then($this->expectCallableOnce());
        $ret->then(function (Sender $sender) {

        });

        $this->loop->run();
    }

    public function testInvalidSender()
    {
        $ret = $this->factory->createSender('some.host.invalid');

        $this->assertInstanceOf('React\Promise\PromiseInterface', $ret);

        $ret->then(null, $this->expectCallableOnce());
        $ret->then(function (Exception $exception) {

        });

        $this->loop->run();
    }
}


src/Sender.php
<?php

namespace Clue\React\Wol;

use Socket\React\Datagram\Socket;
use InvalidArgumentException;

class Sender
{
    private $socket;

    public function __construct(Socket $socket)
    {
        $this->socket = $socket;
        $this->socket->pause();
    }

    public function send($mac)
    {
        $mac = $this->coerceMac($mac);

        $message = "\xFF\xFF\xFF\xFF\xFF\xFF" . str_repeat($this->formatMac($mac), 16);

        $this->socket->send($message);
    }

    /**
     *
     * @param string $mac
     *            mixed case mac address with colon, hyphen or no separators
     * @return string uppercase mac address with colon separators (e.g. 00:11:22:33:44:55)
     * @throws InvalidArgumentException
     */
    public function coerceMac($mac)
    {
        if (strlen($mac) === 12) {
            // no separators => add colons in between
            $mac = implode(':', str_split($mac, 2));
        } elseif (strpos($mac, '-') !== false) {
            // hyphen separators => replace with colons
            $mac = str_replace('-', ':', $mac);
        }
        $mac = strtoupper($mac);

        if (!preg_match('/^(?:[A-F0-9]{2}\:){5}[A-F0-9]{2}$/', $mac)) {
            throw new InvalidArgumentException('Invalid mac address given');
        }

        return $mac;
    }

    private function formatMac($mac)
    {
        $address = '';

        foreach (explode(':', $mac) as $part) {
            $address .= chr(hexdec($part));
        }

        return $address;
    }
}


Wol.php
<?php

use Evenement\EventEmitter;
use React\EventLoop\LoopInterface;
use Socket\React\Datagram\Factory as DatagramFactory;
use InvalidArgumentException;

class Wol extends EventEmitter
{
    private $socket;
    private $address = '255.255.255.255:7';

    public function __construct(LoopInterface $loop)
    {
        $factory = new DatagramFactory($loop);
        $this->socket = $factory->createUdp4();
        $this->socket->setOptionBroadcast();

        $this->socket->pause();
    }

    public function send($mac)
    {
        $mac = $this->coerceMac($mac);

        $message = "\xFF\xFF\xFF\xFF\xFF\xFF" . str_repeat($this->formatMac($mac), 16);

        $this->socket->send($message, $this->address);
    }

    /**
     *
     * @param string $mac
     *            mixed case mac address with colon, hyphen or no separators
     * @return string uppercase mac address with colon separators (e.g. 00:11:22:33:44:55)
     * @throws InvalidArgumentException
     */
    public function coerceMac($mac)
    {
        if (strlen($mac) === 12) {
            // no separators => add colons in between
            $mac = implode(':', str_split($mac, 2));
        } elseif (strpos($mac, '-') !== false) {
            // hyphen separators => replace with colons
            $mac = str_replace('-', ':', $mac);
        }
        $mac = strtoupper($mac);

        if (!preg_match('/(?:[A-F0-9]{2}\:){5}[A-F0-9]{2}/', $mac)) {
            throw new InvalidArgumentException('Invalid mac address given');
        }

        return $mac;
    }

    private function formatMac($mac)
    {
        $address = '';

        foreach (explode(':', $mac) as $part) {
            $address .= chr(hexdec($part));
        }

        return $address;
    }
}
